### PR TITLE
WIP: User Query to power Chart in explore

### PR DESCRIPTION
**Pull request by hughhhh**
### SUMMARY

Updating the `explore/form_data` endpoints to accept `datasource_type = query`. This is one of the main PRs to allow charts to be powered by other datasources.

### TESTING INSTRUCTIONS

To test you'd need to basically create POST + PUT call matching these test:
https://github.com/preset-io/superset/pull/216/files#diff-b60e3914926aca2bf64b704edd1b1ec2e83d11580411d212bc402759f085cf3bR445-R478

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/explore/form_data/api.py [hunk 1]
```diff
@@ -247,7 +248,9 @@ def get(self, key: str) -> Response:
               $ref: '#/components/responses/500'
         """
         try:
-            args = CommandParameters(actor=g.user, key=key)
+            args = CommandParameters(
+                datasource_type=DatasourceType.TABLE, actor=g.user, key=key
```
Comment (hughhhh) [inline superset/explore/form_data/api.py:252]: change these to take datasource_type and cast to enum

File: superset/explore/form_data/schemas.py [hunk 2]
```diff
@@ -33,6 +33,12 @@ class FormDataPostSchema(Schema):
     form_data = fields.String(
         required=True, allow_none=False, description="Any type of JSON supported text."
     )
+    datasource_id = fields.Integer(
+        required=True, allow_none=False, description="Datasource Id."
+    )
+    datasource_type = fields.String(
```
Comment (eschutho) [inline superset/explore/form_data/schemas.py:39]: do you want to add validation choices on this?

Comment (eschutho) [reply superset/explore/form_data/schemas.py:39]: I have some examples on my pr for reference.

File: tests/integration_tests/explore/form_data/api_tests.py [hunk 3]
```diff
@@ -39,6 +44,42 @@
...
 
+@pytest.fixture
+def get_query_datasource():
+    with app.app_context() as ctx:
+        session: Session = ctx.app.appbuilder.get_session
+        engine = session.get_bind()
+        Query.metadata.create_all(engine)  # pylint: disable=no-member
+        query_obj = Query(
+            client_id="foo",
+            database=Database(database_name="my_database", sqlalchemy_uri="sqlite://"),
```
Comment (eschutho) [inline tests/integration_tests/explore/form_data/api_tests.py:55]: can you use `database = get_example_database()` for this?

File: tests/integration_tests/explore/form_data/api_tests.py [hunk 4]
```diff
@@ -392,11 +433,34 @@ def test_delete_not_owner(client, chart_id: int, datasource: SqlaTable, admin_id
...
+def test_post_with_query(client, chart_id: int, get_query_datasource: Query):
+    login(client, "admin")
+    form_data = {
+        "adhoc_filters": [],
+        "viz_type": "sankey",
+        "groupby": ["target"],
+        "metric": "sum__value",
+        "row_limit": 5000,
+        "slice_id": chart_id,
+        "time_range_endpoints": ["inclusive", "exclusive"],
```
Comment (eschutho) [inline tests/integration_tests/explore/form_data/api_tests.py:456]: didn't we remove these time_range_endpoints?

**Comment by eschutho** [comment]
Great. I didn't realize that it would be this simple now to save a chart with a query after just adding in the datasource_id and datasource_type values. Do you still need these other values for the query model in this exploremixin here for it to work? https://github.com/apache/superset/pull/20014/files

**Pull request closed without merging**